AirSkyBoat/AirSkyBoat
# Implemented the Brigand's Chart Quest

**_I affirm:_**


## Please enter a player-facing description
- Implemented the Brigand's Chart quest in Buburimu Peninsula. (HeavensSword)
- Updated Message IDs for fishing in Buburimu Peninsula and Valkurm Dunes. (HeavensSword)
- Updated the Pirate's Chart ??? default message in Valkurm Dunes. (HeavensSword)



## What does this pull request do? (Please be technical)
- Created a full implementation of the Brigand's Chart mini quest using IF.
- Added Chest and quest handling to fishingutils.h/.cpp
- Created new DB table for fishable chests. (Gave some support for implementing custom fished chests)
- Fixed the default fishing message ID offsets for Buburimu and Valkurm
- Added default action for qm1 in Buburimu.
- Added logic for spawning Jade Etui (fishable chests) and adding appropriate quest treasure.
- Added logic for spawning the correct Mob (Puffer Pugil) During quest
- Handle despawning chests and quest cleanup.



## Steps to test these changes
- !additem 1873 (brigands chart)
- !zone {Buburimu Peninsula}
- !gotoname qm1
- Trade the brigand's chart to the ???
- Choose to start the quest
- Note the lvl 20 restriction placed
- Begin fishing (make sure to have a rod and bait/lure equipped. Any)
- You have 100% chance to catch either a Jade Etui or Puffer Pugil with very easy catch difficulty.
- If you catch the Puffer Pugil and keep it spawned, you will not catch another.
- Catch the Etuis and open them to receive your potential loot
NOTE: The 4th chest has a 30% chance to drop the Penguin Ring, while the 5th (final) chest has a 60% chance.
NOTE: The penguin ring is always accompanied by 3 Yellow Globes.

- Note the spawned Hume NPC will give the correct dialog which is spaced out 30 seconds each after the initial "My Penguin Ring..." dialog
NOTE: The entire event is 3 mins long
- The NPC will despawn (and the event ends) 10 seconds after he gives the "I'm fading" message

NOTE: When the event ends, the NPC will despawn, along with any Jade Etuis that have not been opened yet.

Additional Notes: 
- You cannot start this quest if in an alliance (you will be asked to dissolve the alliance). 
- You can start this quest in a party, but only you will receive the lvl 20 restriction.
- You are the only one who will be able to fish up Jade Etuis and the quest Puffer Pugil.
- You are the only one that can open the Jade Etuis.
- Party members can kill/hold the Puffer Pugil for you.






## Special Deployment Considerations
- Update the Puffer Pugil used for this quests level to 10:

`UPDATE fishing_mob SET level=10 WHERE mobid=17261040;`

- Then you need to add the new table "fishing_chest".
NOTE: Run the following or use the newly added "fishing_chest.sql"

```
DROP TABLE IF EXISTS `fishing_chest`;
CREATE TABLE IF NOT EXISTS `fishing_chest` (
  `chestid` int(10) unsigned NOT NULL,
  `name` varchar(64) NOT NULL DEFAULT '',
  `model_index` tinyint(3) unsigned NOT NULL DEFAULT 0,
  `quest_only` tinyint(2) unsigned NOT NULL DEFAULT 0,
  `quest` tinyint(3) unsigned NOT NULL DEFAULT 255,
  `log` tinyint(3) unsigned NOT NULL DEFAULT 255,
  `zoneid` smallint(5) unsigned NOT NULL,
  `areaid` smallint(5) unsigned NOT NULL DEFAULT 0,
  `min_respawn` int(10) unsigned NOT NULL DEFAULT 0,
  `max_respawn` int(10) unsigned NOT NULL DEFAULT 0,
  `difficulty` tinyint(3) unsigned NOT NULL,
  `base_delay` tinyint(2) unsigned NOT NULL,
  `base_move` tinyint(2) unsigned NOT NULL,
  `required_baitid` smallint(5) unsigned NOT NULL DEFAULT 0,
  `alternative_baitid` smallint(5) unsigned NOT NULL DEFAULT 0,
  `required_keyitem` smallint(5) unsigned NOT NULL DEFAULT 0,
  `required_charVar` varchar(64) DEFAULT NULL,
  `required_charVarValue` int(10) unsigned NOT NULL DEFAULT 0,
  `disabled` tinyint(2) unsigned NOT NULL DEFAULT 0,
  PRIMARY KEY (`chestid`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb3 COLLATE=utf8mb3_general_ci;

INSERT INTO `fishing_chest` (`chestid`, `name`, `model_index`, `quest_only`, `quest`, `log`, `zoneid`, `areaid`, `min_respawn`, `max_respawn`, `difficulty`, `base_delay`, `base_move`, `required_baitid`, `alternative_baitid`, `required_keyitem`, `required_charVar`, `required_charVarValue`, `disabled`) VALUES
	(17261035, 'Jade Etui', 2, 1, 255, 255, 118, 0, 0, 0, 10, 10, 15, 0, 0, 0, 'HQuest[BrigChartQuest]Prog', 1, 0),
	(17261036, 'Jade Etui', 2, 1, 255, 255, 118, 0, 0, 0, 10, 10, 15, 0, 0, 0, 'HQuest[BrigChartQuest]Prog', 2, 0),
	(17261037, 'Jade Etui', 2, 1, 255, 255, 118, 0, 0, 0, 10, 10, 15, 0, 0, 0, 'HQuest[BrigChartQuest]Prog', 3, 0),
	(17261038, 'Jade Etui', 2, 1, 255, 255, 118, 0, 0, 0, 10, 10, 15, 0, 0, 0, 'HQuest[BrigChartQuest]Prog', 4, 0),
	(17261039, 'Jade Etui', 2, 1, 255, 255, 118, 0, 0, 0, 10, 10, 15, 0, 0, 0, 'HQuest[BrigChartQuest]Prog', 5, 0);

```
